plugins/mentions/plugin.js

```diff
@@ -0,0 +1,350 @@
+		 *
+		 * @property {String} [template=CKEDITOR.plugins.autocomplete.view.itemTemplate]
+		 * @readonly
+		 */
+		this.template = config.template;
+
+		/**
+		 * {@link CKEDITOR.plugins.autocomplete Autocomplete} instance used by mentions feature to implement autocompletion logic.
+		 *
+		 * @property {CKEDITOR.plugins.autocomplete}
+		 * @private
+		 */
+		this._autocomplete = new CKEDITOR.plugins.autocomplete( editor,
+			getTextTestCallback( this.marker, this.minChars ),
+			getDataCallback( feed, this.marker, this.caseSensitive ) );
+
+		if ( this.template ) {
+			this.changeViewTemplate( this.template );
+		}
+	}
+
+	Mentions.prototype = {
+
+		/**
+		 * Destroys the mentions instance.
+		 * View element and event listeners will be removed from the DOM.
+		 */
+		destroy: function() {
+			this._autocomplete.destroy();
+		},
+
+		/**
+		 * Changes the view template. The given template will be used to render matches in the dropdown.
+		 *
+		 * @param {String} template
+		 */
+		changeViewTemplate: function( template ) {
+			this._autocomplete.view.itemTemplate = new CKEDITOR.template( template );
+		}
+	};
+
+	function getTextTestCallback( marker, minChars ) {
+		var matchPattern = createPattern();
+
+		return function( range ) {
+			if ( !range.collapsed ) {
+				return null;
+			}
+
+			return CKEDITOR.plugins.textMatch.match( range, matchCallback );
+		};
+
+		function matchCallback( text, offset ) {
+			var match = text.slice( 0, offset ).match( matchPattern );
+
+			if ( !match ) {
+				return null;
+			}
+
+			return { start: match.index, end: offset };
+		}
+
+		function createPattern() {
+			var pattern = '[' + marker + ']\\w';
+
+			if ( minChars ) {
+				pattern += '{' + minChars + ',}';
+			} else {
+				pattern += '*';
+			}
+
+			pattern += '$';
+
+			return new RegExp( pattern );
+		}
+	}
+
+	function getDataCallback( feed, marker, caseSensitive ) {
+		return function( query, range, callback ) {
+			// We are removing marker here to give clean query result for the endpoint callback.
+			if ( marker ) {
+				query = query.replace( marker, '' );
+			}
+
+			callback = getDataCallbackWithMarkedItems( marker, callback );
+
+			// Feed is a simple array e.g. [ 'Anna', 'John', 'Thomas' ].
+			if ( CKEDITOR.tools.array.isArray( feed ) ) {
+				var data = indexFeed( feed ).filter( function( item ) {
+					var itemName = item.name;
+
+					if ( !caseSensitive ) {
+						itemName = itemName.toLowerCase();
+						query = query.toLowerCase();
+					}
+
+					return itemName.indexOf( query ) === 0;
+				} );
+
+				callback( data );
+			}
+
+			// Feed is a URL to be requested for a JSON of matches e.g. `/user-controller/get-list/{encodedQuery}`.
+			else if ( typeof feed === 'string' ) {
+				var encodedQueryRegex = new RegExp( ENCODED_QUERY, 'g' ),
+					encodedUrl = feed.replace( encodedQueryRegex, encodeURIComponent( query ) );
+
+				CKEDITOR.ajax.load( encodedUrl, function( data ) {
```

Comment:
This code doesn't handle response race conditions. I'll inlcude the full repro case in review summary.